### PR TITLE
feat: add CUDA support

### PR DESCRIPTION
**Pull request by vmx**
This library is now an abstraction for OpenCL and CUDA. It can be compiled
with support for either or both, via the feature flags `opencl` and `cuda`.

There is now a `Device`, which can point to an OpenCL and/or CUDA device.

You should be able to execute OpenCL and CUDA kernels with the same code
without modifications. You would pass in two closures with the same function
body into `Program::run()`.

To create two closures with the correct signature but the same body, you
can use the `define_closures()` helper macro.

So your code would look like this:

    use rust_gpu_tools::{cuda, define_closures, opencl, GPUError, Program};

    let closures = define_closures!(|program: &opencl::Program | &cuda::Program|
     -> Result<Vec<u8>, GPUError> {
        let input = program.create_buffer::<u8>(128)?;
        let output = program.create_buffer::<u8>(128)?;
        let kernel = program.create_kernel(…)?;
        kernel.arg(&input).arg(&output).run()?;
        let mut out = vec![0u8; 128];
        program.read_into_buffer(&output, 0, &mut out)?;
        Ok(out)
    });

    let program = Program::Cuda(cuda::Program::from_cuda(…)?;
    let results = program.run(closures)?;

### CODE REVIEW

**Comment by vmx** [comment]
This PR looks bigger than it is. For easier review:

 - the `src/device.rs` is similar to `opencl/mod.rs`. Diffing between current `HEAD` of `opencl/mod.rs` and this `src/device.rs` should give a better idea of the changes.
 - the code in the `cuda` directory is very similar to the code in the `opencl` directory, you might also want to diff between those two locally.

**Comment by vmx** [comment]
It might also help to see how it will be used. Here's the WIP in `bellperson` that uses this commit: https://github.com/filecoin-project/bellperson/commit/6958c61856fdcf74da3c0aca4f8c019eea1609fc?w=1

File: Cargo.toml [hunk 1]
```diff
@@ -9,12 +9,18 @@ license = "MIT/Apache-2.0"
 repository = "https://github.com/filecoin-project/rust-gpu-tools"
 
 # See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html
+[features]
+default = ["opencl"]
```
Comment (dignifiedquire) [inline Cargo.toml:13]: I think we can default to both

Comment (vmx) [reply Cargo.toml:13]: The reason I didn't make it the default is that you cannot run the test without having CUDA installed. Although it's true for OpenCL as well, you can easily have OpenCL installed, CUDA is harder to install.

But it's easy enough to disable it, so I guess enabling it by default makes sense.

**Review by dignifiedquire — COMMENTED** [review]
some preliminary thoughts, not through yet

File: src/cuda/mod.rs [hunk 2]
```diff
@@ -0,0 +1,229 @@
...
+use rustacuda::memory::AsyncCopyDestination;
+
+use crate::device::{DeviceUuid, PciId, Vendor};
+use crate::error::{GPUError, GPUResult};
+
+pub struct Buffer<T> {
+    // We cannot use `T` directly for the `DeviceBuffer` as `AsyncCopyDestination` is only
+    // implemented for `u8`.
+    buffer: rustacuda::memory::DeviceBuffer<u8>,
+    length: usize,
```
[the remark this answers is not in the document]
Comment (vmx) [reply src/cuda/mod.rs:15]: That's done in the "document public API PR" which comes once this is merged.

File: src/cuda/mod.rs [hunk 3]
```diff
@@ -0,0 +1,229 @@
...
+    _phantom: std::marker::PhantomData<T>,
+}
+
+#[derive(Debug, Clone)]
+pub struct Device {
+    vendor: Vendor,
+    name: String,
+    memory: u64,
```
[the remark this answers is not in the document]
Comment (vmx) [reply src/cuda/mod.rs:23]: That's done in the "document public API PR" which comes once this is merged.

**Comment by vmx** [comment]
Current state of this PR: It needs more work. The current API is unsafe, but not marked as such. rust-gpu-tools should be a save wrapper, which will need changes on the overall API.

**Comment by vmx** [comment]
This PR is ready for another round of reviews. Things that changed since last time:

 - Operations are mostly sync now, this makes the code simpler and doesn't need any hacks, while it doesn't make any performance difference. I did run the `multiexmultiexp::gpu_multiexp_consistency` test (also with a higher number of elements) and the difference was within the range of differences between runs of the async code.
 - `create_buffer` is now `unsafe`, as it really is. Though there is a new method called `create_buffer_from_slice()`, which is a safe alternative for most of the cases

Things missing before this can be merged:

 - https://github.com/bheisler/RustaCUDA/issues/58 would be great to have, as it makes it possible to include embed kernels directly into the library.
 - I'd also like https://github.com/kenba/opencl3/pull/27 to be merged and released as it is the right thing to have (it would also make the OpenCL need to have a mutable buffer for copying data to the GPU.

**Comment by vmx** [comment]
This one is ready for another round of review. I'd squash it into a single commit once the review it is done.

File: src/cuda/mod.rs [hunk 4]
```diff
@@ -0,0 +1,437 @@
...
+
+        // Transmuting types is safe as long a sizes match.
+        let bytes = unsafe {
+            std::slice::from_raw_parts(slice.as_ptr() as *const T as *const u8, bytes_len)
+        };
+
+        // It is only unsafe as long as the buffer isn't initialized, but that's what we do next.
+        let mut buffer = unsafe { DeviceBuffer::<u8>::uninitialized(bytes_len)? };
+        // It is safe as we synchronize the stream after the call.
+        unsafe { buffer.async_copy_from(bytes, &self.stream)? };
```
Comment (dignifiedquire) [inline src/cuda/mod.rs:196]: is this guranteed to intialize all bytes in `bytes`, independent of `self.stream`?

File: src/cuda/mod.rs [hunk 5]
```diff
@@ -0,0 +1,437 @@
...
+        let result = fun(self, arg);
+        Self::pop_context();
+        result
+    }
+
+    /// Pop the current context.
+    ///
+    /// It panics it it cannot as it's an unrecoverable error.
+    fn pop_context() {
+        rustacuda::context::ContextStack::pop().expect("Cannot remove newly created context.");
```
Comment (dignifiedquire) [inline src/cuda/mod.rs:284]: why is it "newly created"?

File: src/cuda/mod.rs [hunk 6]
```diff
@@ -0,0 +1,437 @@
...
+    }
+}
+
+// TODO vmx 2021-07-07: Check if RustaCUDA types use in `Program` can be made `Send`, so that
+// this manual `Send` implementation is no longer needed.
```
Comment (dignifiedquire) [inline src/cuda/mod.rs:289]: what's the state of this?

File: src/cuda/mod.rs (hunk 4 above)
Comment (vmx) [reply src/cuda/mod.rs:196]: I don't understand what you mean. It transfers all bytes from `bytes` into the memory allocated on the GPU via `buffer`. If the buffer is the same size as the bytes copied, all bytes are initialized.

File: src/cuda/mod.rs (hunk 5 above)
Comment (vmx) [reply src/cuda/mod.rs:284]: Good question :)

File: src/cuda/mod.rs (hunk 4 above)
Comment (dignifiedquire) [reply src/cuda/mod.rs:196]: I didn‘t understand the code right, thanks all good

**Review by dignifiedquire — APPROVED** [review]

**Pull request merged**